File: stack/text_editor/text_editor.py
from collections import deque


class TextEditor:  # LeetCode Q.2296.
    def __init__(self):
        self.cursor_left_chars = []  # Stack based. Chars at cursor's left.
        self.cursor_right_chars = deque()  # Queue based. Chars at cursor's right.

    def add_text(self, text: str) -> None:
        for char in text:
            self.cursor_left_chars.append(char)

    def delete_text(self, k: int) -> int:
        deletions = min(k, len(self.cursor_left_chars))
        for _ in range(deletions):
            self.cursor_left_chars.pop(-1)
        return deletions

    def move_cursor_left(self, k: int) -> str:
        while k > 0 and self.cursor_left_chars:
            char = self.cursor_left_chars.pop(-1)
            self.cursor_right_chars.appendleft(char)
            k -= 1

        str_min_idx = -min(10, len(self.cursor_left_chars))  # Backward idx.
        return "".join(self.cursor_left_chars[str_min_idx:])

    def move_cursor_right(self, k: int) -> str:
        while k > 0 and self.cursor_right_chars:
            char = self.cursor_right_chars.popleft()
            self.cursor_left_chars.append(char)
            k -= 1

        str_min_idx = -min(10, len(self.cursor_left_chars))  # Backward idx.
        return "".join(self.cursor_left_chars[str_min_idx:])

File: stack/text_editor/text_editor.js
const TRACE_LIMIT = 10; // Number of chars to trace to the cursor left.

var TextEditor = function () {
    this.cursor_left = []; // Stack.
    this.cursor_right = []; // Reverse queue: queue's head is actually at idx -1.
};

/** 
 * @param {string} text
 * @return {void}
 */
TextEditor.prototype.addText = function (text) {
    for (const char of text)
        this.cursor_left.push(char);
};

/** 
 * @param {number} k
 * @return {number}
 */
TextEditor.prototype.deleteText = function (k) {
    let deletions = 0;
    while (Math.min(this.cursor_left.length, k) > 0) {
        this.cursor_left.pop();
        deletions++;
        k--;
    }

    return deletions;
};

/** 
 * @param {number} k
 * @return {string}
 */
TextEditor.prototype.cursorLeft = function (k) {
    while (Math.min(this.cursor_left.length, k) > 0) {
        this.cursor_right.push(this.cursor_left.pop());
        k--;
    }

    let cursor_left_str = "";
    let start_idx = -Math.min(this.cursor_left.length, TRACE_LIMIT);
    for (idx = start_idx; idx <= -1; idx++)
        cursor_left_str += this.cursor_left.at(idx);

    return cursor_left_str;
};

/** 
 * @param {number} k
 * @return {string}
 */
TextEditor.prototype.cursorRight = function (k) {
    while (Math.min(this.cursor_right.length, k) > 0) {
        this.cursor_left.push(this.cursor_right.pop());
        k--;
    }

    let cursor_left_str = "";
    let start_idx = -Math.min(this.cursor_left.length, TRACE_LIMIT);
    for (idx = start_idx; idx <= -1; idx++)
        cursor_left_str += this.cursor_left.at(idx);

    return cursor_left_str;
};